Add Custom+Partial Iteration for Particle Scatter and Gather Operations

aliemen opened the pull request:
This pull request implements the following changes to resolve issue [#326](https://github.com/IPPL-framework/ippl/issues/326):

1. **Scatter Operation Enhancements**:
   - Modified `ParticleAttrib::scatter` to accept a custom range policy and hashing array, like `ippl::hash_type`.
   - Added a check to ensure the custom policy matches the number of available elements.
   - Introduced the usage of `mapped_idx` instead of `idx` for particle scattering.
   - Overloaded the non-class `scatter` interface to preserve old functionality while allowing custom range policies.

2. **Gather**:
   - Modified `ParticleAttrib::gather` to include a boolean parameter that determines whether to add to or overwrite the attribute's values.
   - Changed the non-class `gather` interface to maintain backward compatibility.

**Testing**:
- Added a new test file `test/particle/TestHashedScatter.cpp`.
- Modified the existing `TestGather` to validate the new functionality (use "+=" instead of "+").


closes #326

srikrrish commented on src/Particle/ParticleAttrib.hpp:
```diff
@@ -205,14 +218,24 @@ namespace ippl {
+        attrib.scatter(f, pp, policy_type(0, attrib.getParticleCount())); 
+    }
+
+    // Second implementation for custom range policy, but without index array
```
A bit more explanation on this would be helpful.

srikrrish reviewed (changes requested):
@aliemen I have only very minor comments like adding/expanding documentation. Otherwise it looks good. You have tests. Is it also possible to convert them to unit tests just like the ones for scatter and gather?

aliemen commented:
@srikrrish I added comments/documentation for the scatter/gather functions. I added them for the implementation of the `ParticleAttrib` class as well as for the non in-class interfaces. Please let me know if you think that is too much, but I think it makes sense that way.

I haven't done this yet, but I will also try to convert the test file into two unit tests.

aliemen commented:
@srikrrish I now also pushed the unit tests. Since it was not a lot that was missing, I also added tests for the old/simple `scatter`/`gather`. I tested it on multiple ranks (1, 2, 4), on CPU as well as on GPU. In the following is a bit of explanation regarding the tests.

## Dimension Independence

You can simply uncomment the parameters with different dimensions under "`TestTypes`" to test for dimensions higher than 3. However, it obviously gets very memory hungry/slow when creating the fields. I tested it for a mesh size of 8 for up to 5 dimensions and this worked fine for alle tests.

At the moment, I only added 1, 2 and 3 dimensions with a grid size of 128 (such that it runs reasonably quick for the usually important dimension values). 

## What Tests Did I Add?

Here is a list of the 5 new unit tests:
1. `GatherTestReplace`. This one tests the normal (old) `gather` functionality. It initializes particles with charge `10` and a constant field with values `1.0`. After the `gather`, the charges should all be replaced with "`1.0`" (which is asserted for each particle).
2. `GatherTestIncrement`. This tests the flag `addToAttribute = true`. It initializes the charges and the field to `1.0`. After the `gather`, all charges should then have the value `2.0` (their old value plus the field value).
3. `ScatterSimpleTest`. This tests the normal (old) `scatter` functionality. It simply initializes charges with `1.0`, scatters them on a field (initialized to `0`) and then compares the total charges.
4. `ScatterCustomRangeTest`. Scatters only the first "`nLoc / 2 + rank`" particles using a custom `Kokkos::RangePolicy`, then tests the correctness like under 3).
5. `ScatterCustomHashTest`. Generates a `hash_type` of "`nLoc / 2 + rank`" random indices, scatters only those indices onto the field then tests the correctness like under 3). The random indices are chosen by shuffling an index array `[1, ..., nLoc]` and selecting the first `nLoc / 2 + rank` indices.
